leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/BaseCertificateVerifier.java

```diff
@@ -76,4 +92,72 @@ protected void validateSubject(final DTLSSession session, final X509Certificate
         throw new HandshakeException(
                 "Certificate chain could not be validated - server identity does not match certificate", alert);
     }
+
+    protected CertPath expandCertPath(CertPath certPath, X509Certificate[] trustedCertificates) {
```

Comment:
This code sounds not so easy to read.
Maybe it deserves to be rewritten.  (lot of break, lot of hidden exception)
Is it something you copy/paste ?

Do we really need this maxDepth check ?

Reply:
I wrote it my self.

Logic is kinda simple -- just find previous cert from trust store and make sure it is the right cert (verify check) and so that it is valid.

Verify check is important when new CA is issued with same name or CA certificate is re-keyed and/or re-issued. It is quite common to start transition period to new CA in controlled fashion. Eg. first distributing new CA's and then when distributed then start changing server certs.

Stop when we found root ca or if some-one made circular chain then protect against that with depth limit.

If everything went well and something was modified -> return modified chain.

If there was an error or no modifications return original chain.

Java crypto library has done most heavy lifting on certificate checks so we can do simple checks here.

Reply:
OK I think this is the last point to address. 🙂

So this code do both thing : 
 - check for chain loop
 - verify certificate signature

(maybe this should be separated in 2 functions)

Tell me if I'm wrong : 
 1. we check loop only for certificate in truststore ?
 2. About verification this should not be done in  `CertPathUtil.validateCertificatePath(truncateCertificatePath, certPath, trustedCertificates);`  ? 

(I'm pretty sure I missed something :thinking:)

Reply:
It reconstructs the missing chain elements from trust store. And part of that process is verifying that correct certificate is picked from trust store.

Reply:
CertPathUtil does not pick certificates properly from trust store if there are multiple CA certs there with same subject DN.

If we make sure that it is correct before going there then it hopefully copes with that.

But that function has:

```java
		CertPathValidator validator = CertPathValidator.getInstance("PKIX");
		PKIXParameters params = new PKIXParameters(trustAnchors);
		// TODO: implement alternative means of revocation checking
		params.setRevocationEnabled(false);
		validator.validate(verifyCertPath, params);
```

Which does the PKI validation magic within Java JRE which is what we want to do as that is rather complex beast.

And this validator is also a picky -- eg. there should not be matching certificates in trust store vs. in chain.

Reply:
OK so this method is to prepare certchain because of kind of limitation of `CertPathUtil.validateCertificatePath` ?
I vaguely remember that Achim talk about it previously. Maybe it's time to invoke him. 
![invocation](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fmedia.tenor.com%2Fimages%2F8e1349237fb7b0636a48a55ff92e906d%2Ftenor.gif&f=1&nofb=1)
@boaks what do you think about this ?

Reply:
I passed `SecurityTest` without `expandCertPath` to see if we have some tests affected by this "issue".
And [registered_device_with_x509cert_to_server_with_x509cert_rootca_certificate_ca_domain_root_ca_given](https://github.com/eclipse/leshan/pull/923/commits/4351e658843ab4885e6a9978dee46d08101eeebf#diff-7c035814cdb5eaaf7e1414ea6beb0249d697c66dbc99a2ce669fdb5660b2c5deR766) doesn't pass.
I tried with cf-2.5.0-SNAPSHOT and test doesn't pass too. So this is probably not what Achim was talking about.

Reply:
Just a note -- later we need also the full chain when we are doing policy based X.509 certificate device auto registration and this can also be the place for that path re-construction.

Reply:
You didn't answer about : 
 1. we check loop only for certificate in truststore ?

Reply:
Thx for the detailed explanation.

Let me try to explain what I would say by : "we check loop only for certificate in truststore ?"

I talk about depth guard.
I understand that  : 
1. we start to check by the end of the initial certpath. `X509Certificate cert = chain.get(chain.size() - 1);`
2. then we only add certificate from `trustedCertificate` (chain.add(caCert);)

So if we discover a loop, it seems to me that it was in the truststore?

Reply:
So, the strategy seems to complete the path up to the root.

For trusting intermediate CAs, that requires still to truncate that path before the PKIX CertPathValidator.
Assuming multiple CA-Certificates with the same DN requires then a redesign in `CertPathUtil.validateCertificatePath`. So, wait for more info about that ...

Reply:
> So if we discover a loop, it seems to me that it was in the truststore?

I was probably thinking originally the whole chain in here. 

But you are true -- the certificate loop would be in trust store.

Now the question would be how it would end up there and shall we have a safe guard in place.

Thinking aloud for possible scenarios for certificate loop:

- If you would have trusted certs from Mozilla or so I would expect them to do also good job in this sense so that would not be an vector.
- One could inject corporate CA's to servers -- not a problem for (embedded) client.
- One could use UI of device to configure additional trusts -- out-of-scope.
- When performing bootstrap one can inject new Server Identity and that would be dropped to trust store -> compromised server -- In this case it probably gets sorted out and clients need to have normal ways to get back in (eg. not get stuck).
- When performing CoAP-EST's CA certs request -- same compromise scenario as above.
- Someone hacked the device -- bigger problem already -- out-of-scope.
- Some security tester proofing a point

So I suppose only scenarios we have here is:
- sanity check that we don't accidentally go to infinite loop. I have never seen 32 deep or even 8 deep CA chains. So value should be safe bet.
- server compromise and making sure that feet of devices can automatically get back in. You probably need to issue new certs in this kind of scenario triggering clients to connect again to bootstrap server and hope that you didn't need to change trust point for bootstrap.